   ---
title: "compare DGN predictions hapmap2 v wtccc genotyped"
author: "Heather E. Wheeler"
date: "`r Sys.time()`"
output: html_document
---

```{r}
date <- Sys.Date()
"%&%" = function(a,b) paste(a,b,sep="")
library(ggplot2)
library(dplyr)
library(tidyr)

my.dir = '/Volumes//im-lab/nas40t2/hwheeler/PrediXcan_CV/GTEx_2014-06013_release/transfers/PrediXmod/DGN-WB/DGN-calc-weights/'

lassoHap<-read.table(my.dir %&% 'DGN-WB_exp_10-foldCV_elasticNet_alpha1_imputedSNPs_chr1-22_2015-02-02.txt',header=T)
enHap<-read.table(my.dir %&% 'DGN-WB_exp_10-foldCV_elasticNet_alpha0.5_imputedSNPs_chr1-22_2015-02-02.txt',header=T)
psHap<-read.table(my.dir %&% 'DGN-WB_exp_10-foldCV_polyscore_Pthresh0.001_imputedSNPs_chr1-22_2015-02-09.txt',header=T)

lassoWT<-read.table(my.dir %&% 'DGN-WB_exp_10-foldCV_elasticNet_alpha1_wtcccGenotypedSNPs_chr1-22_2015-03-12.txt',header=T)
enWT<-read.table(my.dir %&% 'DGN-WB_exp_10-foldCV_elasticNet_alpha0.5_wtcccGenotypedSNPs_chr1-22_2015-03-12.txt',header=T)

en1KG<-read.table(my.dir %&% 'DGN-WB_exp_10-foldCV_elasticNet_alpha0.5_1000G_chr1-22_2015-04-02.txt',header=T)

lassoHap<-lassoHap[complete.cases(lassoHap),]
lassoWT<-lassoWT[complete.cases(lassoWT),]
enHap<-enHap[complete.cases(enHap),]
enWT<-enWT[complete.cases(enWT),]
psHap<-psHap[complete.cases(psHap),]
en1KG<-en1KG[complete.cases(en1KG),]
dim(lassoHap)
dim(lassoWT)
dim(enHap)
dim(enWT)
dim(psHap)
dim(en1KG)

lasso<-inner_join(lassoHap,lassoWT,by='gene')
dim(lasso)
summary(lm(R2.y~R2.x,lasso))
p<-ggplot(lasso,aes(x=R2.x,y=R2.y)) +geom_point()+ geom_abline(intercept=0, slope=1,color='red') + ggtitle("DGN-WB Cross-Validation")
p1<-p + xlab(expression("Hapmap2 SNPs LASSO R"^2)) + ylab(expression("WTCCC genotyped SNPs LASSO R"^2)) + theme_bw(20)
print(p1)

png(file="DGN-CV_LASSO.wtccc_v_hapmap.png")
p1
dev.off()

en<-inner_join(enHap,enWT,by='gene')
dim(en)
summary(lm(R2.y~R2.x,en))
p<-ggplot(en,aes(x=R2.x,y=R2.y)) +geom_point()+ geom_abline(intercept=0, slope=1,color='red') + ggtitle("DGN-WB Cross-Validation")
p1<-p + xlab(expression("Hapmap2 SNPs Elastic Net R"^2)) + ylab(expression("WTCCC genotyped SNPs Elastic Net R"^2)) + theme_bw(20)
print(p1)

png(file="DGN-CV_ENalpha0.5.wtccc_v_hapmap.png")
p1
dev.off()

ps<-inner_join(psHap,enWT,by='gene')
dim(ps)
summary(lm(CV.R2~R2,ps))
p<-ggplot(ps,aes(x=CV.R2,y=R2)) +geom_point()+ geom_abline(intercept=0, slope=1,color='red') + ggtitle("DGN-WB Cross-Validation")
p1<-p + xlab(expression("Hapmap2 SNPs Polyscore (p<1e-04) R"^2)) + ylab(expression("WTCCC genotyped SNPs Elastic Net R"^2)) + theme_bw(20)
print(p1)

png(file="DGN-CV_ENalpha0.5.wtccc_v_Polyscore1e-4hapmap.png")
p1
dev.off()

en<-inner_join(enHap,en1KG,by='gene')
dim(en)
summary(lm(R2.y~R2.x,en))
p<-ggplot(en,aes(x=R2.x,y=R2.y)) +geom_point()+ geom_abline(intercept=0, slope=1,color='red') + ggtitle("DGN-WB Cross-Validation")
p1<-p + xlab(expression("Hapmap2 SNPs Elastic Net R"^2)) + ylab(expression("1000G SNPs Elastic Net R"^2)) + theme_bw(20)
print(p1)

png(file="DGN-CV_ENalpha0.5.1000G_v_hapmap.png")
p1
dev.off()

###calc number of genes with 1000G R2 > HapMap2 R2
diffen<-abs(en$R2.y-en$R2.x)>0.10
table(diffen)
diffen<-abs(en$R2.y-en$R2.x)>0.05
table(diffen)
diffen<-abs(en$R2.y-en$R2.x)>0.01
table(diffen)

diven<-en[abs(en$R2.y-en$R2.x)>0.10,]
diven <-  select(diven,gene,R2.x,R2.y)
colnames(diven) <- c('gene','hapmap.R2','1000.R2')
diven
```

